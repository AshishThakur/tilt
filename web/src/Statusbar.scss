@import "constants";

.Statusbar {
  position: fixed;
  left: 0;
  right: 0;
  bottom: 0;
  z-index: $z-sidebar;
  height: $statusbar-height;
  background-color: $color-white;
  color: $color-gray;
  display: flex;
  align-items: stretch;
  border-top: 1px solid $color-gray-darkest;
  border-bottom: 1px solid $color-gray-darkest;
  box-sizing: border-box;
}

.Statusbar-panel {
  display: flex;
  align-items: center;
  box-sizing: border-box;
}
.Statusbar-panel + .Statusbar-panel {
  border-left: 1px dashed $color-gray-lightest;
}

// Errors and Warnings
.err-warn {
  display: flex;
  align-items: stretch;
}
.err-warn-item {
  display: flex;
  align-items: center;
  padding-left: $spacing-unit / 2;
  padding-right: $spacing-unit / 2;
  // Eyeball to make them the same width, and not change size when contents update
  min-width: $spacing-unit * 3.5;
}
.err-warn-item + .err-warn-item {
  border-left: 1px solid rgba($color-gray-lightest, $translucent);
}
.err-warn .count {
  font-weight: bold;
}
.err-warn .icon {
  fill: $color-gray-lightest;
  opacity: $translucent;
  margin-right: $spacing-unit / 3;
  filter: drop-shadow(-1px 1px 0px rgba($color-gray-darkest, $translucent-ish));
}
.err-warn .icon--error {
  fill: $color-red;
  opacity: 1;
  filter: none;
}
.err-warn .icon--warning {
  color: $color-yellow;
  opacity: 1;
  filter: none;
}

// Status Message
.statusMsg {
  flex-grow: 1;
  display: flex;
  justify-content: space-between;
  padding-left: $spacing-unit / 2;
  padding-right: $spacing-unit / 2;
}

.statusMsg-item--last-edit .label {
  text-transform: uppercase;
  font-size: $font-size-small;
  font-weight: bold;
  color: $color-gray-lightest;
  margin-right: $spacing-unit / 4;
}

// Progress
.progress {
  width: $sidebar-width;
  max-width: $sidebar-maxWidth;
  justify-content: space-between;
  display: flex;
  padding-left: $spacing-unit / 2;
  padding-right: $spacing-unit / 2;
}
.progress .current {
  font-weight: bold;
}
