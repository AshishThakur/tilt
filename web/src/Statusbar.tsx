import React, { PureComponent } from "react"
import { ReactComponent as LogoSvg } from "./assets/svg/logo.svg"
import { ReactComponent as ErrorSvg } from "./assets/svg/error.svg"
import { ReactComponent as WarningSvg } from "./assets/svg/warning.svg"
import { combinedStatus, warnings } from "./status"
import "./Statusbar.scss"
import { combinedStatusMessage } from "./combinedStatusMessage"
import { Build } from "./types"
import mostRecentBuildToDisplay from "./mostRecentBuild"

class StatusItem {
  public warningCount: number = 0
  public up: boolean = false
  public hasError: boolean = false
  public name: string
  public currentBuild: Build
  public lastBuild: Build | null
  public podStatus: string
  public pendingBuildSince: string
  public buildHistory: Array<Build>
  public pendingBuildEdits: Array<string>

  /**
   * Create a pared down StatusItem from a ResourceView
   */
  constructor(res: any) {
    this.name = res.Name
    this.warningCount = warnings(res).length

    let status = combinedStatus(res)
    this.up = status == "ok"
    this.hasError = status == "error"
    this.currentBuild = res.CurrentBuild
    this.buildHistory = res.BuildHistory
    this.lastBuild = res.BuildHistory ? res.BuildHistory[0] : null
    this.podStatus = res.ResourceInfo && res.ResourceInfo.PodStatus
    this.pendingBuildSince = res.PendingBuildSince
    this.pendingBuildEdits = res.PendingBuildEdits
  }
}

type StatusBarProps = {
  items: Array<StatusItem>
}

class Statusbar extends PureComponent<StatusBarProps> {
  errorWarningPanel(errorCount: number, warningCount: number) {
    return (
      <section className="Statusbar-panel err-warn">
        <div className="err-warn-item err-warn-item--error">
          <ErrorSvg className={`icon ${errorCount > 0 ? "icon--error" : ""}`} />
          <p>
            <span className="count">{errorCount}</span> error
            {errorCount === 1 ? "" : "s"}
          </p>
        </div>
        <div className="err-warn-item">
          <WarningSvg
            className={`icon ${warningCount > 0 ? "icon--warning" : ""}`}
          />
          <p>
            <span className="count">{warningCount}</span> warning
            {warningCount === 1 ? "" : "s"}
          </p>
        </div>
      </section>
    )
  }

  progressPanel(upCount: number, itemCount: number) {
    return (
      <section className="Statusbar-panel progress">
        <p>
          <span className="current">{upCount}</span>/{itemCount} running
        </p>
        <LogoSvg className="icon--logo" />
      </section>
    )
  }

  statusMessagePanel(build: any, editMessage: string) {
    return (
      <section className="Statusbar-panel statusMsg">
        <p className="statusMsg-item">
          {combinedStatusMessage(this.props.items)}
        </p>
        <p className="statusMsg-item statusMsg-item--last-edit">
          <span className="label">Last Edit:</span>
          <span className="file">{build ? editMessage : "—"}</span>
        </p>
      </section>
    )
  }

  render() {
    let errorCount = 0
    let warningCount = 0
    let upCount = 0

    let items = this.props.items
    items.forEach(item => {
      if (item.hasError) {
        errorCount++
      }
      if (item.up) {
        upCount++
      }
      warningCount += item.warningCount
    })
    let errorWarningPanel = this.errorWarningPanel(errorCount, warningCount)

    let build = mostRecentBuildToDisplay(this.props.items)
    let editMessage = ""
    if (build && build.edits.length > 0) {
      editMessage = `${build.name} ‣ ${build.edits[0]}`
      if (build.edits.length > 1) {
        editMessage += `[+${build.edits.length - 1} more]`
      }
    }
    let statusMessagePanel = this.statusMessagePanel(build, editMessage)

    let resCount = items.length
    let progressPanel = this.progressPanel(upCount, resCount)

    return (
      <div className="Statusbar">
        {errorWarningPanel}
        {statusMessagePanel}
        {progressPanel}
      </div>
    )
  }
}

export default Statusbar

export { StatusItem }
