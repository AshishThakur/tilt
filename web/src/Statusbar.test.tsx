import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer';
import Statusbar from './Statusbar'

it('renders without crashing if theres no last build', () => {
  const tree = renderer
    .create(<Statusbar items={[]} toggleSidebar={null} />)
    .toJSON()

  expect(tree).toMatchSnapshot()
})
