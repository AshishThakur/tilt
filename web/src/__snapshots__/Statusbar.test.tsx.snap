// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing if theres no last build 1`] = `
<div
  className="Statusbar"
>
  <div
    className="Statusbar-panel Statusbar-panel--error"
  >
    <span
      aria-label="Error"
      className="icon"
      role="img"
    >
       
    </span>
     
    <span>
      0
       
      Errors
    </span>
  </div>
  <div
    className="Statusbar-panel Statusbar-panel--warning"
  >
    <span
      aria-label="Warning"
      className="icon"
      role="img"
    >
       
    </span>
     
    <span>
      0
       
      Warnings
    </span>
  </div>
  <div
    className="Statusbar-panel Statusbar-panel--spacer"
  >
     
  </div>
  <button
    className="Statusbar-panel Statusbar-panel--up"
    onClick={null}
  >
    0
     / 
    0
     resources up 
    <svg
      className="icon"
    >
      logo-imagemark.svg
    </svg>
  </button>
</div>
`;
